File: src/github_service.js
let OctokitModule;

const API_VERSION = '2022-11-28';

function toDate(value) {
    if (!value) return new Date(0);
    const date = new Date(value);
    return Number.isNaN(date.getTime()) ? new Date(0) : date;
}

function loginOf(user) {
    if (!user) return null;
    return user.login || user.name || null;
}

function activityDate(activity) {
    if (!activity) return new Date(0);
    return toDate(activity.submitted_at || activity.created_at || activity.updated_at);
}

function isSameUser(user, username) {
    const login = loginOf(user);
    if (!login || !username) return false;
    return login.toLowerCase() === username.toLowerCase();
}

function isSeen(seenPRs, url, activityAt) {
    if (!seenPRs || !url) return false;
    const marker = seenPRs[url];
    if (!marker) return false;

    if (typeof marker === 'string') {
        return toDate(activityAt).getTime() <= toDate(marker).getTime();
    }
    if (marker.seenAt) {
        return toDate(activityAt).getTime() <= toDate(marker.seenAt).getTime();
    }
    return false;
}

function sortNewestFirst(events) {
    return events.sort((a, b) => b.date.getTime() - a.date.getTime());
}

class GitHubService {
    constructor(token) {
        this.token = token;
        this.octokit = null;
        this.username = null;
        this.lastRateLimit = null;
    }

    async getOctokit() {
        if (!this.octokit) {
            if (!OctokitModule) {
                const mod = await import('octokit');
                OctokitModule = mod.Octokit;
            }
            this.octokit = new OctokitModule({ auth: this.token });
        }
        return this.octokit;
    }

    async verifyUser() {
        try {
            const octokit = await this.getOctokit();
            const { data: user } = await octokit.rest.users.getAuthenticated();
            this.username = user.login;
            return user;
        } catch (error) {
            console.error('Auth verification error:', error.status || error.message);
            return null;
        }
    }

    rememberRateLimit(response) {
        const headers = response?.headers || {};
        if (headers['x-ratelimit-remaining'] !== undefined) {
            this.lastRateLimit = {
                remaining: Number(headers['x-ratelimit-remaining']),
                resetAt: headers['x-ratelimit-reset']
                    ? new Date(Number(headers['x-ratelimit-reset']) * 1000).toISOString()
                    : null,
            };
        }
    }

    async safeFetch(fn, params) {
        try {
            if (typeof fn !== 'function') return [];
            const response = await fn({ ...params, per_page: 100 });
            this.rememberRateLimit(response);
            return Array.isArray(response?.data) ? response.data : [];
        } catch (e) {
            return [];
        }
    }

    async searchPullRequests(username) {
        const octokit = await this.getOctokit();
        const params = {
            q: `is:pr is:open involves:${username}`,
            per_page: 100,
            headers: { 'X-GitHub-Api-Version': API_VERSION },
        };
        if (typeof octokit.request === 'function') {
            const response = await octokit.request('GET /search/issues', params);
            this.rememberRateLimit(response);
            return response.data?.items || [];
        }
        if (octokit.rest?.search?.issuesAndPullRequests) {
            const response = await octokit.rest.search.issuesAndPullRequests(params);
            this.rememberRateLimit(response);
            return response.data?.items || [];
        }
        return [];
    }

    async getPullRequestDiff(owner, repo, pull_number) {
        try {
            const octokit = await this.getOctokit();
            const response = await octokit.request('GET /repos/{owner}/{repo}/pulls/{pull_number}', {
                owner,
                repo,
                pull_number,
                headers: {
                    accept: 'application/vnd.github.v3.diff',
                    'X-GitHub-Api-Version': API_VERSION
                }
            });
            this.rememberRateLimit(response);
            return typeof response.data === 'string' ? response.data : JSON.stringify(response.data);
        } catch (error) {
            console.error(`Error fetching diff for PR #${pull_number}:`, error.message);
            return '';
        }
    }

    async getStatus(seenPRs = {}, includeSeen = false) {
        try {
            const octokit = await this.getOctokit();
            if (!this.username) {
                const user = await this.verifyUser();
                if (!user) return null;
            }

            const username = this.username;
            const alerts = {
                merge_conflicts: [],
                review_required: [],
                re_review_needed: [],
                my_pr_activity: [],
                meta: { rateLimit: this.lastRateLimit, generatedAt: new Date().toISOString() },
            };
            const items = await this.searchPullRequests(username);

            for (const item of items) {
                const parts = String(item.repository_url || '').split('/').filter(Boolean);
                const repo = parts.pop();
                const owner = parts.pop();
                if (!owner || !repo || !item.number) continue;

                let pr;
                try {
                    const pullResponse = await octokit.rest.pulls.get({
                        owner,
                        repo,
                        pull_number: item.number,
                        headers: { 'X-GitHub-Api-Version': API_VERSION },
                    });
                    this.rememberRateLimit(pullResponse);
                    pr = pullResponse.data;
                } catch (e) {
                    console.error(`Error fetching PR #${item.number}:`, e.message);
                    continue;
                }

                const base = { owner, repo, pull_number: item.number };

                const [reviews, reviewComments, issueComments] = await Promise.all([
                    this.safeFetch(octokit.rest.pulls.listReviews ? octokit.rest.pulls.listReviews.bind(octokit.rest.pulls) : null, base),
                    this.safeFetch(octokit.rest.pulls.listReviewComments ? octokit.rest.pulls.listReviewComments.bind(octokit.rest.pulls) : null, base),
                    this.safeFetch(octokit.rest.issues?.listComments ? octokit.rest.issues.listComments.bind(octokit.rest.issues) : null, {
                        owner, repo, issue_number: item.number,
                    }),
                ]);

                const events = [];
                for (const review of reviews) {
                    if (review && review.state && review.state !== 'PENDING') {
                        events.push({
                            type: 'review',
                            state: review.state,
                            user: loginOf(review.user) || 'reviewer',
                            date: activityDate(review),
                            created_at: review.submitted_at || review.created_at,
                            id: review.id,
                            body: review.body || '',
                            headSha: review.commit_id || null,
                        });
                    }
                }
                for (const comment of reviewComments) {
                    if (comment) {
                        events.push({
                            type: 'inline_comment',
                            user: loginOf(comment.user) || 'reviewer',
                            date: activityDate(comment),
                            created_at: comment.created_at,
                            id: comment.id,
                            body: comment.body || '',
                            headSha: comment.commit_id || null,
                        });
                    }
                }
                for (const comment of issueComments) {
                    if (comment) {
                        events.push({
                            type: 'issue_comment',
                            user: loginOf(comment.user) || 'usuario',
                            date: activityDate(comment),
                            created_at: comment.created_at,
                            id: comment.id,
                            body: comment.body || '',
                            headSha: null,
                        });
                    }
                }
                sortNewestFirst(events);

                const mine = sortNewestFirst(events.filter(e => isSameUser({ login: e.user }, username)));
                const others = sortNewestFirst(events.filter(e => !isSameUser({ login: e.user }, username)));
                const latestMine = mine[0] || null;
                const latestOther = others[0] || null;
                const headSha = pr.head?.sha || null;
                const latestActivityAt = events[0]?.date || toDate(pr.updated_at);
                const seen = isSeen(seenPRs, pr.html_url || item.html_url, latestActivityAt);

                const common = {
                    title: pr.title || item.title || `PR #${item.number}`,
                    url: pr.html_url || item.html_url,
                    updated_at: pr.updated_at,
                    head_sha: headSha,
                    number: item.number,
                    repository: `${owner}/${repo}`,
                    author: loginOf(pr.user) || username,
                    latest_activity_at: latestActivityAt.toISOString(),
                    has_conflict: pr.mergeable === false || pr.mergeable_state === 'dirty',
                };

                // Conflictos de Merge
                if (pr.mergeable === false || pr.mergeable_state === 'dirty') {
                    if (includeSeen || !seen) {
                        alerts.merge_conflicts.push({
                            ...common,
                            state: '💥 Conflictos de Merge (Bloqueado)',
                        });
                    }
                }

                // CASO C: Mi propio PR
                if (isSameUser(pr.user, username)) {
                    const otherFormal = others.find(e => e.type === 'review' && ['CHANGES_REQUESTED', 'APPROVED'].includes(e.state));
                    const newOther = latestOther && (!latestMine || latestOther.date > latestMine.date);
                    
                    let detail = null;
                    if (otherFormal?.state === 'CHANGES_REQUESTED') {
                        detail = `Cambios pedidos por @${otherFormal.user} ⚠️`;
                    } else if (otherFormal?.state === 'APPROVED') {
                        detail = `Aprobado por @${otherFormal.user} ✅`;
                    } else if (newOther) {
                        detail = `Comentario de @${latestOther.user} 💬`;
                    } else if (others.length > 0) {
                        detail = `Actividad reciente en tu PR 💬`;
                    } else {
                        detail = `Abierto y esperando revisión ⏳`;
                    }

                    if (includeSeen || !seen) {
                        alerts.my_pr_activity.push({
                            ...common,
                            state: detail,
                        });
                    }
                    continue;
                }

                // CASOS A y B: PRs de otros
                const requested = (pr.requested_reviewers || []).some(r => isSameUser(r, username));
                if (requested && !latestMine) {
                    if (includeSeen || !seen) {
                        alerts.review_required.push({
                            ...common,
                            state: 'Revisión Requerida: Asignado ⏳',
                        });
                    }
                    continue;
                }

                if (latestMine) {
                    const newCommits = Boolean(headSha && latestMine.headSha && headSha !== latestMine.headSha)
                        || Boolean(toDate(pr.updated_at) > latestMine.date);
                    const newOtherActivity = latestOther && latestOther.date > latestMine.date;

                    if ((newCommits || newOtherActivity) && (includeSeen || !seen)) {
                        let detail = 'Commits nuevos 🔄';
                        if (newOtherActivity && !newCommits) detail = `Respuesta de @${latestOther.user} 💬`;
                        if (newOtherActivity && newCommits) detail = `Commits o respuesta de @${latestOther.user} 💬`;
                        alerts.re_review_needed.push({
                            ...common,
                            state: `Re-revisión Pendiente: ${detail}`,
                        });
                    }
                }
            }

            for (const key of ['merge_conflicts', 'review_required', 're_review_needed', 'my_pr_activity']) {
                alerts[key].sort((a, b) => toDate(b.latest_activity_at || b.updated_at) - toDate(a.latest_activity_at || a.updated_at));
            }
            alerts.meta.rateLimit = this.lastRateLimit;
            return alerts;
        } catch (error) {
            console.error('Error fetching GitHub PR status:', error.message);
            return null;
        }
    }
}

module.exports = GitHubService;

File: github_pet/github_service.py
"""GitHub REST client and deterministic PR alert classifier."""
from datetime import datetime, timezone
import json
import os
from typing import Any, Callable, Optional
from urllib.error import HTTPError, URLError
from urllib.parse import quote
from urllib.request import Request, urlopen

from .models import (
    MY_PR_ACTIVITY, RE_REVIEW_NEEDED, REVIEW_REQUIRED,
    Alert, MonitorSnapshot, PullRequest,
)

Json = dict[str, Any]


def parse_dt(value: Optional[str]) -> Optional[datetime]:
    if not value:
        return None
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


class GitHubError(RuntimeError):
    pass


class GitHubClient:
    def __init__(self, token: Optional[str] = None, api_base: str = "https://api.github.com",
                 opener: Callable[..., Any] = urlopen):
        self.token = token or os.getenv("GITHUB_TOKEN") or os.getenv("GH_TOKEN")
        self.api_base = api_base.rstrip("/")
        self.opener = opener

    def get(self, path: str, params: Optional[dict[str, str]] = None) -> Any:
        query = ""
        if params:
            query = "?" + "&".join(f"{quote(k)}={quote(v)}" for k, v in params.items())
        headers = {"Accept": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        request = Request(self.api_base + path + query, headers=headers)
        try:
            with self.opener(request, timeout=20) as response:
                return json.loads(response.read().decode("utf-8"))
        except (HTTPError, URLError, TimeoutError) as exc:
            raise GitHubError(f"GitHub API request failed: {exc}") from exc

    def authenticated_user(self) -> Json:
        return self.get("/user")

    def assigned_prs(self, username: str) -> list[Json]:
        return self.get("/search/issues", {"q": f"is:pr is:open review-requested:{username}", "per_page": "100"})["items"]

    def authored_prs(self, username: str) -> list[Json]:
        return self.get("/search/issues", {"q": f"is:pr is:open author:{username}", "per_page": "100"})["items"]

    def pr_details(self, owner: str, repo: str, number: int) -> Json:
        return self.get(f"/repos/{quote(owner)}/{quote(repo)}/pulls/{number}")

    def reviews(self, owner: str, repo: str, number: int) -> list[Json]:
        return self.get(f"/repos/{quote(owner)}/{quote(repo)}/pulls/{number}/reviews", {"per_page": "100"})

    def issue_comments(self, owner: str, repo: str, number: int) -> list[Json]:
        return self.get(f"/repos/{quote(owner)}/{quote(repo)}/issues/{number}/comments", {"per_page": "100"})

    def review_comments(self, owner: str, repo: str, number: int) -> list[Json]:
        return self.get(f"/repos/{quote(owner)}/{quote(repo)}/pulls/{number}/comments", {"per_page": "100"})

    def to_pr(self, item: Json, username: str, details: Optional[Json] = None,
              reviews: Optional[list[Json]] = None, issue_comments: Optional[list[Json]] = None,
              review_comments: Optional[list[Json]] = None, previous_head_sha: Optional[str] = None,
              last_seen_at: Optional[datetime] = None) -> PullRequest:
        d = details or item
        repo_value = d.get("base", {}).get("repo", {}).get("full_name")
        if repo_value:
            owner, repo = repo_value.split("/", 1)
        else:
            parts = item.get("repository_url", "").rstrip("/").split("/")
            owner, repo = (parts[-2], parts[-1]) if len(parts) >= 2 else ("", "")
        all_reviews = reviews or []
        mine = [r for r in all_reviews if r.get("user", {}).get("login", "").lower() == username.lower()]
        latest = max((parse_dt(r.get("submitted_at")) for r in mine if r.get("submitted_at")), default=None)
        latest_mine = max(mine, key=lambda r: parse_dt(r.get("submitted_at")) or datetime.min) if mine else None
        state = latest_mine.get("state") if latest_mine else None
        approvals = sum(r.get("state") == "APPROVED" for r in all_reviews)
        others = [r for r in all_reviews if r.get("user", {}).get("login", "").lower() != username.lower()]
        other_approvals = sum(r.get("state") == "APPROVED" for r in others)
        other_requested_changes = sum(r.get("state") == "CHANGES_REQUESTED" for r in others)
        issue_comments = issue_comments or []
        review_comments = review_comments or []
        activity_times = [parse_dt(x.get("created_at")) for x in issue_comments + review_comments]
        activity_times += [parse_dt(x.get("submitted_at")) for x in all_reviews]
        latest_activity = max((x for x in activity_times if x), default=None)
        def is_new(value: Optional[str]) -> bool:
            timestamp = parse_dt(value)
            return bool(timestamp and (last_seen_at is None or timestamp > last_seen_at))
        new_issue_comments = sum(is_new(x.get("created_at")) for x in issue_comments)
        new_review_comments = sum(is_new(x.get("created_at")) for x in review_comments)
        new_commits = bool(previous_head_sha and d.get("head", {}).get("sha") != previous_head_sha)
        activity_is_new = None if last_seen_at is None else bool(latest_activity and latest_activity > last_seen_at)
        return PullRequest(
            number=int(d["number"]), title=d["title"], url=d["html_url"], owner=owner, repo=repo,
            author=d.get("user", {}).get("login", ""), updated_at=parse_dt(d.get("updated_at")) or datetime.now(timezone.utc),
            head_sha=d.get("head", {}).get("sha", ""), reviewers=tuple(x.get("login", "") for x in d.get("requested_reviewers", [])),
            my_review_state=state, latest_review_at=latest, comments_count=int(d.get("comments", 0)),
            review_comments_count=int(d.get("review_comments", 0)), requested_changes=other_requested_changes > 0, approvals=approvals,
            other_approvals=other_approvals, other_requested_changes=other_requested_changes,
            new_issue_comments=new_issue_comments, new_review_comments=new_review_comments,
            new_commits=new_commits, activity_is_new=activity_is_new, latest_activity_at=latest_activity)


def classify_prs(prs: list[PullRequest], username: str, now: Optional[datetime] = None) -> MonitorSnapshot:
    """Classify PRs. Priority: own-PR activity, re-review, review assignment."""
    alerts: list[Alert] = []
    for pr in prs:
        if pr.author.lower() == username.lower():
            collaborator_activity = pr.other_approvals or pr.other_requested_changes or pr.new_issue_comments or pr.new_review_comments
            legacy_activity = pr.activity_is_new is None and (pr.approvals or pr.comments_count or pr.review_comments_count)
            if pr.activity_is_new is not False and (pr.requested_changes or collaborator_activity or legacy_activity):
                if pr.other_requested_changes or pr.requested_changes:
                    detail = "Cambios pedidos ⚠️"
                elif pr.other_approvals or pr.approvals:
                    detail = "Aprobación de un colaborador ✅"
                else:
                    detail = "Comentario nuevo 💬"
                alerts.append(Alert(MY_PR_ACTIVITY, "Actividad en tu PR", pr, f"⚡ Actividad en tus PRs: {pr.title} ({detail})"))
        elif username.lower() in {r.lower() for r in pr.reviewers} and not pr.my_review_state:
            alerts.append(Alert(REVIEW_REQUIRED, "Review pendiente", pr, f"⏳ Revisión Requerida: {pr.title} (Asignado)"))
        elif pr.latest_review_at and pr.updated_at > pr.latest_review_at:
            if pr.new_commits or pr.new_issue_comments or pr.new_review_comments or pr.activity_is_new is True:
                alerts.append(Alert(RE_REVIEW_NEEDED, "Re-revisión", pr, f"🔄 Re-revisión Pendiente: {pr.title} (Commits o respuesta de @{pr.author} 💬)"))
    return MonitorSnapshot(tuple(alerts), now or datetime.now(timezone.utc))
